### PR TITLE
fix(clickhouse): add support for DateTime64(precision, tz) and Tuples()

### PR DESCRIPTION
**Pull request by ogirardot**
### Brief summary of the change made
The changes introduce support on the Clickhouse dialect for types and type casts related to `DateTime64` and `Tuple` (structs in clickhouse), the following syntaxes were unparseable using sqlfluff 3.1.0 (and dbt-clickhouse 1.8.1, with `sqlfluff-templater-dbt==3.1.0` as example :
```
SELECT (1, 'two', '2024-01-01')::Tuple(id Int64, name String, created_at Date32) as struct;
-- and 
SELECT '2024-01-01'::DateTime64(3, 'Europe/Paris') as datetime;
```

### Are there any other side effects of this change that we should be aware of?
Nope

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).

### CODE REVIEW

File: test/fixtures/dialects/clickhouse/create_table.yml [hunk 1]
```diff
@@ -507,11 +507,10 @@ file:
...
-            bracketed_arguments:
-              bracketed:
-                start_bracket: (
-                data_type_identifier: String
-                end_bracket: )
+          bracketed:
+            start_bracket: (
+            word: String
+            end_bracket: )
       - end_bracket: )
```
Comment (alanmcruickshank) [inline test/fixtures/dialects/clickhouse/create_table.yml:514]: This change looks like a regression. While it won't be a `bracketed_arguments` anymore, it should still be part of the `data_type` segment. Can you make sure your change supports `Nullable` types appropriately?

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Thanks for contributing this!

Changes look mostly good. Thanks for following the same pattern used in the bigquery dialect.

One thing which isn't critical, but would be nice is whether we can get `DateTime64` to parse as a `data_type_identifier` rather than a `keyword` in the test cases 🤔 . That way they'll get handled appropriately by the capitalisation rules.

I _think_ the best way to achieve that would be instead of just `"DATETIME64"`, you use the slightly longer `StringParser("DATETIME64", DatatypeIdentifierSegment)`. It's a big more verbose but I think it's more robust on how things get treated in the rules.

File: test/fixtures/dialects/clickhouse/create_table.yml (hunk 1 above)
Comment (ogirardot) [reply test/fixtures/dialects/clickhouse/create_table.yml:514]: oh you're right, I'll look into it

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 2]
```diff
@@ -328,15 +328,66 @@ class BracketedArguments(ansi.BracketedArguments):
...
+    the target type in casts.
+    """
+
+    type = "data_type"
+    match_grammar = OneOf(
+        Ref("TupleTypeSegment"),
+        Ref("DatatypeIdentifierSegment"),
+        Ref("NumericLiteralSegment"),
+        Sequence(
+            OneOf("DATETIME64"),
```
[the remark this answers is not in the document]
Comment (ogirardot) [reply src/sqlfluff/dialects/dialect_clickhouse.py]: Got it using StringParser and CodeSegment

File: test/fixtures/dialects/clickhouse/create_table.yml (hunk 1 above)
Comment (ogirardot) [reply test/fixtures/dialects/clickhouse/create_table.yml]: I decided to handle the nullable case more explicitely than before

**Comment by ogirardot** [comment]
@alanmcruickshank it should be good to go, I fixed the last violations of the pre-commit. Thank you for your time and review

**Review by alanmcruickshank — APPROVED** [review]
LGTM! 👍

Thanks for your contribution, and especially for taking the time to make the edits.

**Pull request merged**